core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java

```diff
@@ -72,20 +96,64 @@ public FilteredQueryBuilder queryName(String queryName) {
         return this;
     }
 
+    /** Returns the name of this query. */
+    public String queryName() {
+        return queryName;
+    }
+
     @Override
-    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject(NAME);
-        if (queryBuilder != null) {
-            builder.field("query");
-            queryBuilder.toXContent(builder, params);
+    protected boolean doEquals(FilteredQueryBuilder other) {
+        return Objects.equals(queryBuilder, other.queryBuilder) &&
+                Objects.equals(filterBuilder, other.filterBuilder) &&
+                Objects.equals(boost, other.boost) &&
+                Objects.equals(queryName, other.queryName);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(queryBuilder, filterBuilder, boost, queryName);
+    }
+
+    @Override
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+
+        Query query = (queryBuilder == null) ? null : queryBuilder.toQuery(parseContext);
```

Comment:
I believe I didn't change the logic (other than making sure that any generated query gets added to named queries, and after this review round gets the boost applied) - would be great to get one final check from @jpountz though as I changed the way the final Lucene query is generated quite a bit though.

> As for the behaviour, is it correct that none of the inner query is mandatory at the moment 
> (before the refactoring) and that doesn't change here? That means match_all query right?

Nearly right: Originally in the parser we started with a match_all query. We than parsed both, the filter and query part. If either returned a null value, we ended up returning null as the final query.

The match_all somewhat-default wasn't (and isn't) built into the FilteredQueryBuilder though: Users always had to got through the constructor that required to specify both, query and filter but allowed both to be null. 

Now trying to get the knot out of my head.